### PR TITLE
add is_supergreedy() to linear extensions

### LINKED ISSUE #24700: Add is_supergreedy() to linear extension
<div id="comment:0"></div>

Add a function to check if a linear extension of a poset is supergreedy. (Compare to #24632.)

Component: **combinatorics**

Author: **Rohan Garg**

Reviewer: **Jori Mäntysalo**

_Issue created by migration from https://trac.sagemath.org/ticket/24700_

### PR DESCRIPTION
**Pull request by Sandstorm831**
This is my pull request originally created on trac, which I thought pushing on github.
Fixes #24700

### CODE REVIEW

**Comment by Sandstorm831** [comment]
Hello everyone, I have created this pull request in view of changes committed by me in the open issue #24700 , I had a positive review on it. Please tell if anything else needs to be done to merge this pull request.

File: src/sage/combinat/posets/linear_extensions.py [hunk 1]
```diff
@@ -40,7 +40,7 @@
 
 
 class LinearExtensionOfPoset(ClonableArray,
-        metaclass=InheritComparisonClasscallMetaclass):
+                             metaclass=InheritComparisonClasscallMetaclass):
```
Comment (dimpase) [inline src/sage/combinat/posets/linear_extensions.py]: looks like empty space there? can it be removed?

Comment (dimpase) [reply src/sage/combinat/posets/linear_extensions.py]: Ah, OK, no problem, sorry.

Comment (Sandstorm831) [reply src/sage/combinat/posets/linear_extensions.py]: yes, I will remove it, please tell if there are any more edits for it.

**Review by dimpase — APPROVED** [review]
OK

File: src/sage/combinat/posets/linear_extensions.py (hunk 1 above)
Comment (dimpase) [reply src/sage/combinat/posets/linear_extensions.py]: never mind, it's OK

**Comment by dimpase** [comment]
for now, we keep the rule that only the release manager merges positively reviewed PRs.

**Comment by Sandstorm831** [comment]
No problem, I will wait. ThankYou @dimpase for the information

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
There are code style issues (see failing github check). Can you please fix them?

**Comment by Sandstorm831** [comment]
Sir, can you tell what are the changes that I have to do.

**Comment by dimpase** [comment]
Click on "Details" for the failing Lint/Code style check. You'll see https://github.com/sagemath/sage/actions/runs/4123554420/jobs/7121712300 pointing to what you have to fix (trailing whitespace in few lines in your code here)

**Comment by Sandstorm831** [comment]
I have fixed the codestyle issues and now the test is passing, please tell if any other change needs to be done

File: src/sage/combinat/posets/linear_extensions.py [hunk 2]
```diff
@@ -252,6 +252,70 @@ def is_greedy(self):
...
+    def is_supergreedy(self):
+        r""""
+        Return ``True`` if the linear extension is supergreedy.
+
+        A linear extension `[e_1, e_2, \ldots, e_n]` is *supergreedy* if
+        for every `i`, either if there is a minimal element `e_{i+1}` in
+        `[e_{i+1}, \ldots, e_n]` which is in the upper cover of `e_j` in
+        `[e_1, \ldots, e_i]` for which j is maximum or if no such element
+        exist `e_{i+1}` is the any of the minimal element in
+        `[e_i+1,\dots, e_n]`.
```
Comment (tobiasdiez) [inline src/sage/combinat/posets/linear_extensions.py]: I think there is a typo here.

Is this what you wanted to say?
```suggestion
        `[e_1, \ldots, e_i]` for which j is maximum or, if no such element
        exist, `e_{i+1}` is a minimal element in
        `[e_i+1,\dots, e_n]`.
```

Comment (Sandstorm831) [reply src/sage/combinat/posets/linear_extensions.py]: No, it isn't a typo, I intentionally changed it to imply the definition of supergreedy linear extensions more firmly

Comment (tobiasdiez) [reply src/sage/combinat/posets/linear_extensions.py]: Oh, I misread your text. Sorry, nevermind.

**Review by tobiasdiez — APPROVED** [review]

**Comment by dimpase** [comment]
sorry,  I don't understand what "the any of ..." means.

**Comment by dimpase** [comment]
do you mean "any of" ? Unless "any" has a special meaning, e.g. you have a function called "any()", so that it becomes a noun (normally it is an adjective) one cannot put an article in front of it.

**Comment by dimpase** [comment]
should it be " any of the minimal elements" ?

(drop the article, pluralise "element")

**Comment by Sandstorm831** [comment]
Actually any of means, if there is no maximal element than we have to pick up any of the minimal element, which in case of total order set will be 1 only, but in case of partial order it can be many because not all of the elements needs not to be comparable, so if there are more than 1 minimal element, then we can choose any of them.

**Comment by dimpase** [comment]
even more idiomatic English would be "any minimal element".

**Comment by Sandstorm831** [comment]
Yes, It can be rephrased as any minimal in the range, but author had written like that only, so I followed it

**Comment by dimpase** [comment]
Once again, this is very broken English here:
```
        A linear extension `[e_1, e_2, \ldots, e_n]` is *supergreedy* if
        for every `i`, either if there is a minimal element `e_{i+1}` in
        `[e_{i+1}, \ldots, e_n]` which is in the upper cover of `e_j` in
        `[e_1, \ldots, e_i]` for which j is maximum or if no such element
        exist `e_{i+1}` is the any of the minimal element in
        `[e_i+1,\dots, e_n]`.
```

* "`[e_1, \ldots, e_i]` for which j is maximum" - the "which" there, what it is referring to? I don't see.
* "j" there should be "`j`"
* there should be "," after "maximum
* "no such element
        exist `e_{i+1}` is the any of the minimal element in
        `[e_i+1,\dots, e_n]`." makes no sense to me. 

For the latter, could you explain what you are trying to say? Because I don't follow. 
E.g. should it start as "no such element `e_{i+1}` exists" ?

**Comment by Sandstorm831** [comment]
Let me put it in a simpler sense. If there is a linear extension which is supergreedy then, if we take first _i_ elements of it then, _i+1_ element will either be only a minimal element in range _i+1_ to _n_ (if the following condition don't meet)  **or** it will be that minimal element in range _i+1_ to _n_ which is also the upper cover of the element which is maximum among all the elements in range _1_ to _i_ having those minimal elements as their upper cover.

Answer to your queries
- _which_ there refers to the `e_j`element, here `e_j` should be max among those elements in given range which have their upper covers among the minimal elements

- treat `e_{i+1}` as the the next element which will come after `e_{i}`. If the upper cover condition doesn't met then it means all other minimal elements present are not comparable to the elements present upto `i` and thus any of the minimal element can be choosen as next element in the linear extension to be a supergreedy linear extension.

Hope this answers your queries, please let me know for anything else.

**Comment by dimpase** [comment]
please see https://math.stackexchange.com/a/4637076/39797 for a way to write down the definition you need correctly and in a readable way.

**Comment by Sandstorm831** [comment]
Thank You @dimpase, Actually I was trying, not to incorporate complex notation in my definition just like definition of is_greedy(), as a result of which the definition had become confusing. I have followed the following paper in understanding the definition of supergreedy linear extensions https://www.researchgate.net/publication/228026499_Super-Greedy_Linear_Extensions_of_Ordered_Sets.

**Comment by dimpase** [comment]
You cannot use `$` for TeX maths in docstrings. Use backticks instead.

**Comment by dimpase** [comment]
You should add a reference to  https://link.springer.com/article/10.1007/BF00337892 for the definition of supergreedy

**Review by dimpase — CHANGES_REQUESTED** [review]
This is not how you add references to papers. You need to put the textual record (authors, title, etc) in 
`src/doc/en/reference/references/index.rst`, and refer to it using `[...]_` syntax. See e.g. how it is done with `[Stan2009]_` in `src/sage/combinat/posets/linear_extensions.py`

File: src/sage/combinat/posets/linear_extensions.py [hunk 3]
```diff
@@ -252,6 +252,73 @@ def is_greedy(self):
                         return False
         return True
 
+    def is_supergreedy(self):
+        r""""
```
Comment (dimpase) [inline src/sage/combinat/posets/linear_extensions.py]: it seems there should be only 3 `"`, not 4

**Review by dimpase — APPROVED** [review]
OK

File: src/sage/combinat/posets/linear_extensions.py [hunk 4]
```diff
@@ -252,6 +252,73 @@ def is_greedy(self):
...
+        H = P.hasse_diagram()
+
+        def next_elements(H, linext):
+            k = len(linext)
+            S = []
+            while not S:
+                if not k:
+                    S = [x for x in H.sources() if x not in linext]
+                else:
+                    S = [x for x in H.neighbor_out_iterator(linext[k-1]) if x not in linext and all(low in linext for low in H.neighbor_in_iterator(x))]
```
Comment (dimpase) [inline src/sage/combinat/posets/linear_extensions.py]: such a long line should better be split. I wonder whether our linters ought to pick this up, @tobiasdiez ?

Comment (tobiasdiez) [reply src/sage/combinat/posets/linear_extensions.py]: It gives an error above 160 characters, see https://github.com/sagemath/sage/blob/fbb412787f6a762442c41a2d7cd01715d9994cb9/src/tox.ini#L130. There are quite a few instances of long lines already in the code, and it is also not clear to me what an optimal line length should be. For example "Black defaults to 88 characters per line, which happens to be 10% over 80. This number was found to produce significantly shorter files than sticking with 80 (the most popular)"

File: src/sage/combinat/posets/linear_extensions.py [hunk 5]
```diff
@@ -252,6 +252,73 @@ def is_greedy(self):
...
+        TESTS::
+
+            sage: T = Poset()
+            sage: T.linear_extensions()[0].is_supergreedy()
+            True
+        """
+        P = self.poset()
+        H = P.hasse_diagram()
+
+        def next_elements(H, linext):
```
Comment (dimpase) [inline src/sage/combinat/posets/linear_extensions.py]: `H` is not modified, why to you pass it as an argument to `next_elements()` ? 

Besides, I think that  a nested function is only needed when you do recursive calls of it, and/or
if you call it more than in one place in the code. Here you can just put its body where your line 315 
(i.e. where it is called).

**Review by dimpase — CHANGES_REQUESTED** [review]

**Comment by dimpase** [comment]
Are you going to do the modification I proposed?

**Comment by Sandstorm831** [comment]
@dimpase sir, sorry for the delay, but for some reason, my reply to your requested changes was put in pending.

**Comment by tobiasdiez** [comment]
@vbraun It seems something went wrong here. The merge commit https://github.com/sagemath/sage/commit/3b94a2ef9159744d4a99e5183896981b96ce07b5 doesn't include the most recent changes of https://github.com/sagemath/sage/pull/34970/commits/150b9bad9def778b8de0f255a546c2a9b17fa16c

**Comment by Sandstorm831** [comment]
Should I recommit those changes ??

**Comment by dimpase** [comment]
> Should I recommit those changes ??

you need to resolve the merge conflicts and recommit. Not sure why this got messed up, but it should be easy.

**Comment by Sandstorm831** [comment]
@dimpase I have resolved the merge conflict, please have a look.

**Comment by Sandstorm831** [comment]
```sage -t --random-seed=30286532797624533416704009338271799637 sage/combinat/posets/linear_extensions.py
    [174 tests, 0.39 s]```
Tests of the files I have committed changes have passes, but the following has failed
```
File "doc/ca/intro/index.rst", line 492, in doc.ca.intro.index
Failed example:
    solve([x + y == 6, x - y == 4], x, y)
Exception raised:
    Traceback (most recent call last):
      File "/__w/sage/sage/src/sage/doctest/forker.py", line 695, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/__w/sage/sage/src/sage/doctest/forker.py", line 1093, in compile_and_execute
        exec(compiled, globs)
      File "<doctest doc.ca.intro.index[1]>", line 1, in <module>
        solve([x + y == Integer(6), x - y == Integer(4)], x, y)
      File "/__w/sage/sage/src/sage/symbolic/relation.py", line 1137, in solve
        m = maxima(f)
      File "sage/misc/lazy_import.pyx", line 404, in sage.misc.lazy_import.LazyImport.__call__
        return self.get_object()(*args, **kwds)
      File "sage/misc/lazy_import.pyx", line 225, in sage.misc.lazy_import.LazyImport.get_object
        return self._get_object()
      File "sage/misc/lazy_import.pyx", line [26](https://github.com/sagemath/sage/actions/runs/4229657807/jobs/7346221409#step:11:27)1, in sage.misc.lazy_import.LazyImport._get_object
        self._object = getattr(__import__(self._module, {}, {}, [self._name]), self._name)
      File "/__w/sage/sage/src/sage/interfaces/maxima_lib.py", line 119, in <module>
        ecl_eval("(set-pathnames)")
      File "sage/libs/ecl.pyx", line 1350, in sage.libs.ecl.ecl_eval
        cpdef EclObject ecl_eval(str s):
      File "sage/libs/ecl.pyx", line 1373, in sage.libs.ecl.ecl_eval
        o=ecl_safe_eval(python_to_ecl(s, True))
      File "sage/libs/ecl.pyx", line [32](https://github.com/sagemath/sage/actions/runs/4229657807/jobs/7346221409#step:11:33)1, in sage.libs.ecl.ecl_safe_eval
        raise RuntimeError("ECL says: {}".format(message))
    RuntimeError: ECL says: Could not create directory "/github/home/.sage/maxima"
    C library error: "File exists"
**********************************************************************
1 item had failures:
   1 of 175 in doc.ca.intro.index
    [1[36](https://github.com/sagemath/sage/actions/runs/4229657807/jobs/7346221409#step:11:37) tests, 1 failure, 6.[39](https://github.com/sagemath/sage/actions/runs/4229657807/jobs/7346221409#step:11:40) s]
```
And a warning has been displayed, although tests have passed
```
doctest:warning
  File "/usr/lib/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.8/threading.py", line 921, in _bootstrap_inner
    self._started.set()
  File "/usr/lib/python3.8/threading.py", line 528, in set
    self._cond.notify_all()
  File "/usr/lib/python3.8/threading.py", line 371, in notify_all
    self.notify(len(self._waiters))
  File "/usr/lib/python3.8/threading.py", line 354, in notify
    waiters_to_notify = _deque(_islice(all_waiters, n))
  File "/usr/lib/python3.8/warnings.py", line 109, in _showwarnmsg
    sw(msg.message, msg.category, msg.filename, msg.lineno,
:
RuntimeWarning: cypari2 leaked 140469342337792 bytes on the PARI stack
```
What should I do @dimpase ?

**Comment by dimpase** [comment]
the latest test failure is not relevant to this ticket. 

I'd like someone who works on posets to have a look at this on the maths side of this ticket: @mantepse - can you please?

Ideally, to check that this all works using some extra info on supergreedy posets, maybe using FindStat?

**Comment by dimpase** [comment]
I've restarted the failed test.

**Comment by Sandstorm831** [comment]
All tests passed this time

**Comment by mantepse** [comment]
I'll have a look later today.

**Comment by mantepse** [comment]
sorry, this is not ready:

* I think the documentation is misleading, if not wrong: the elements of a supergreedy linear extension are not comparable, as  `[x_1<x_2<...<x_t]` suggests.  Why don't you follow the conventions in `is_greedy`?
* you compute `H.sources()` and `self[:i]` repeatedly, which is unnecessary.
* since `k < i+1`, the construction `self[:i+1][k]` should be the same as `self[k]`.
* line 310 is super hard to read, and it is not clear to me, what it does.
* did you rename the variables from `supergreedy_linear_extensions_iterator` on purpose?

**Comment by Sandstorm831** [comment]
Thank you @mantepse for the feedback. 
- documentation is indeed misleading at that point, and now had corrected it by following the documentation of `is_greedy()`.
- computing `H.sources()` and `self[:i+1]` is indeed unnecessary and I had implemented other optimal approach.
- 3rd point got corrected by addressing 2nd point.
- about line 310 previously also there was a doubt about its length, so I had simplified the function and also reduced its length.
- While making this function I was not aware that there is `supergreedy_linear_extensions_iterator` function, when I got to know about it, I am almost done with my function, As that algorithm was much concise from mine function so I kept my variables and taken the algorithm.
hope this addresses the issues, please tell if there any other changes needs to be done additionally.

**Comment by dimpase** [comment]
yeah, I have no idea how a part of this PR ended up in `develop`.

**Comment by Sandstorm831** [comment]
This time a test timed out of `cone.py`

**Comment by Sandstorm831** [comment]
Should I have to do anything @mantepse

**Comment by mantepse** [comment]
Sorry, I ran out of time.  I'll have another look right now.

**Comment by mantepse** [comment]
* please remove trailing whitespace (in line 258, for example)
* operators should be surrounded by whitespace: `linext=[]` should be `linext = []`, and this assignment should be done just before the for loop
* I think that
```py
            if self[i+1] in L:
                continue
            else:
                return False
```
can be simplified to
```py
            if self[i+1] not in L:
                return False
```
* possibly the `while` loop could be simplified, but this would take some work: currently the code creates a list `L`, and then checks whether it contains `self[i+1]`.  This is somewhat more work than necessary: if we encounter an element `x == self[i+1]` we can stop.  However, I could not find an elegant solution.  The best I could come up with is as follows:
```py
    for i in range(len(self)-2):
        linext.append(self[i])
        for y in reversed(linext):
            L = [x for x in H.neighbor_out_iterator(y)
                 if x not in linext
                 and all(low in linext for low in H.neighbor_in_iterator(x))]
            if L:
                break
        else:
            L = (x for x in sources if x not in linext)
        if self[i+1] not in L:
            return False
```
or, also not creating `L` in the first case,
```py
    for i in range(len(self)-2):
        linext.append(self[i])
        e = self[i+1]
        for y in reversed(linext):
            L = (x for x in H.neighbor_out_iterator(y)
                 if x not in linext
                 and all(low in linext for low in H.neighbor_in_iterator(x)))
            empty = True
            for x in L:
                empty = False
                if x == e:
                    break
            else:
                if not empty:
                    return False
            if not empty:
                break
        else:
            L = (x for x in sources if x not in linext)
            if e not in L:
                return False
```
However, since the lists are short, you should take the simplest version.

**Comment by Sandstorm831** [comment]
@mantepse , I will make changes shortly, but on looking to changes you proposed in last point doesn't seem to have much changes, as in the both the codes, we have to create an list to store the values, if any value enters then as in my algorithm also, while loop automatically exits as in your case, breaks out of for loop. So should I change the code.

As you are saying that when we encounter an element `x == self[i+1]`we break out, then I can implement that, but I don't know if that would create any significant change.

**Comment by mantepse** [comment]
I think you should take the simplest solution. The final one I posted is too complicated. Of the other two, which one is shorter (fewer lines of code)?

**Comment by Sandstorm831** [comment]
Mine code will take 13 lines and yours is taking 12 lines, should I change the code then ?

**Comment by mantepse** [comment]
Whatever you find simpler.

**Comment by Sandstorm831** [comment]
@mantepse, I have incorporated all the changes, on looking closely, your code seems easy to understand then mine, so I have updated the code. Please check and let me know if there are any more changes needs to be incorporated.

**Comment by mantepse** [comment]
One final thing: in the docstring, the definition is not very clear, in particular, the $x$ in $p(x)$ is unclear.  I find the definition in Kierstead and Trotter quite good, it is reproduced at https://www.findstat.org/StatisticsDatabase/St001106.  Maybe you could adapt it for the docstring.

**Comment by Sandstorm831** [comment]
@mantepse , I have improved the definition in the docstrings. Please have a look and let me know if there any more changes needs to be done.

**Comment by Sandstorm831** [comment]
```
2023-02-26T12:25:06.1327632Z File "doc/ca/intro/index.rst", line 492, in doc.ca.intro.index
2023-02-26T12:25:06.1327885Z Failed example:
2023-02-26T12:25:06.1328173Z     solve([x + y == 6, x - y == 4], x, y)
2023-02-26T12:25:06.1328411Z Exception raised:
2023-02-26T12:25:06.1328636Z     Traceback (most recent call last):
2023-02-26T12:25:06.1328952Z       File "/__w/sage/sage/src/sage/doctest/forker.py", line 695, in _run
2023-02-26T12:25:06.1329287Z         self.compile_and_execute(example, compiler, test.globs)
2023-02-26T12:25:06.1329646Z       File "/__w/sage/sage/src/sage/doctest/forker.py", line 1093, in compile_and_execute
2023-02-26T12:25:06.1329902Z         exec(compiled, globs)
2023-02-26T12:25:06.1330184Z       File "<doctest doc.ca.intro.index[1]>", line 1, in <module>
2023-02-26T12:25:06.1330561Z         solve([x + y == Integer(6), x - y == Integer(4)], x, y)
2023-02-26T12:25:06.1330873Z       File "/__w/sage/sage/src/sage/symbolic/relation.py", line 1137, in solve
2023-02-26T12:25:06.1331116Z         m = maxima(f)
2023-02-26T12:25:06.1331440Z       File "sage/misc/lazy_import.pyx", line 404, in sage.misc.lazy_import.LazyImport.__call__
2023-02-26T12:25:06.1331766Z         return self.get_object()(*args, **kwds)
2023-02-26T12:25:06.1332368Z       File "sage/misc/lazy_import.pyx", line 225, in sage.misc.lazy_import.LazyImport.get_object
2023-02-26T12:25:06.1332802Z         return self._get_object()
2023-02-26T12:25:06.1333144Z       File "sage/misc/lazy_import.pyx", line 261, in sage.misc.lazy_import.LazyImport._get_object
2023-02-26T12:25:06.1333519Z         self._object = getattr(__import__(self._module, {}, {}, [self._name]), self._name)
2023-02-26T12:25:06.1333881Z       File "/__w/sage/sage/src/sage/interfaces/maxima_lib.py", line 119, in <module>
2023-02-26T12:25:06.1334224Z         ecl_eval("(set-pathnames)")
2023-02-26T12:25:06.1334523Z       File "sage/libs/ecl.pyx", line 1350, in sage.libs.ecl.ecl_eval
2023-02-26T12:25:06.1334790Z         cpdef EclObject ecl_eval(str s):
2023-02-26T12:25:06.1335088Z       File "sage/libs/ecl.pyx", line 1373, in sage.libs.ecl.ecl_eval
2023-02-26T12:25:06.1335377Z         o=ecl_safe_eval(python_to_ecl(s, True))
2023-02-26T12:25:06.1335673Z       File "sage/libs/ecl.pyx", line 321, in sage.libs.ecl.ecl_safe_eval
2023-02-26T12:25:06.1335993Z         raise RuntimeError("ECL says: {}".format(message))
2023-02-26T12:25:06.1336378Z     RuntimeError: ECL says: Could not create directory "/github/home/.sage/maxima/binary/5_45_0/ecl/21_2_1"
2023-02-26T12:25:06.1336693Z     C library error: "File exists"
2023-02-26T12:25:11.1458396Z **********************************************************************
2023-02-26T12:25:11.1458869Z 1 item had failures:
2023-02-26T12:25:11.1459101Z    1 of 175 in doc.ca.intro.index
```
This test failed.

File: src/sage/combinat/posets/linear_extensions.py [hunk 6]
```diff
@@ -255,23 +255,23 @@ def is_greedy(self):
...
-        to be a minimal element of `P-\{x_1,...,x_i\}` which
-        maximizes `p`.
+
+        A linear extension of a Poset `P` with elements `\{_1,x_2,...,x_t\}`
+        is *super greedy*, if it can be obtained using the following
+        algorithm: Choose `x_1` to be a minimal element of `P`;
+        suppose `X = \{x_1,...,x_i\}` have been chosen; let `M` be
+        the set of minimal elements of `P/X`. If there is an element
+        of `M` which covers an element `x_j` in `X`, then let `x_{i+1}`
+        be one of these such that `j` is maximal; otherwise, choose `x_{i+1}
```
Comment (mantepse) [inline src/sage/combinat/posets/linear_extensions.py]: There is a backtick (`) missing after x_{i+1}.

File: src/sage/combinat/posets/linear_extensions.py [hunk 7]
```diff
@@ -255,23 +255,23 @@ def is_greedy(self):
...
-        define `p(x)` to be the largest `j\leq i` such that `x_j<x`
-        if such a `j` exists and 0 otherwise; choose `x_{i+1}`
-        to be a minimal element of `P-\{x_1,...,x_i\}` which
-        maximizes `p`.
+
+        A linear extension of a Poset `P` with elements `\{_1,x_2,...,x_t\}`
+        is *super greedy*, if it can be obtained using the following
+        algorithm: Choose `x_1` to be a minimal element of `P`;
+        suppose `X = \{x_1,...,x_i\}` have been chosen; let `M` be
+        the set of minimal elements of `P/X`. If there is an element
```
Comment (mantepse) [inline src/sage/combinat/posets/linear_extensions.py]: It should be $P\setminus X$, (`P\setminus X`) not $P/X$.

File: src/sage/combinat/posets/linear_extensions.py [hunk 8]
```diff
@@ -255,23 +255,23 @@ def is_greedy(self):
...
-        the following procedure: Choose `x_1` to be a minimal
-        element of `P`; suppose `x_1,...,x_i` have been chosen;
-        define `p(x)` to be the largest `j\leq i` such that `x_j<x`
-        if such a `j` exists and 0 otherwise; choose `x_{i+1}`
-        to be a minimal element of `P-\{x_1,...,x_i\}` which
-        maximizes `p`.
+
+        A linear extension of a Poset `P` with elements `\{_1,x_2,...,x_t\}`
+        is *super greedy*, if it can be obtained using the following
+        algorithm: Choose `x_1` to be a minimal element of `P`;
```
Comment (mantepse) [inline src/sage/combinat/posets/linear_extensions.py]: after a colon, English requires lower case.

File: src/sage/combinat/posets/linear_extensions.py [hunk 9]
```diff
@@ -255,23 +255,23 @@ def is_greedy(self):
...
-        A linear extension `[x_1<x_2<...<x_t]` of a finite ordered
-        set `P=(P, <)` is *super greedy* if it can be obtained using
-        the following procedure: Choose `x_1` to be a minimal
-        element of `P`; suppose `x_1,...,x_i` have been chosen;
-        define `p(x)` to be the largest `j\leq i` such that `x_j<x`
-        if such a `j` exists and 0 otherwise; choose `x_{i+1}`
-        to be a minimal element of `P-\{x_1,...,x_i\}` which
-        maximizes `p`.
+
+        A linear extension of a Poset `P` with elements `\{_1,x_2,...,x_t\}`
```
Comment (mantepse) [inline src/sage/combinat/posets/linear_extensions.py]: Poset should be lower case.

Comment (mantepse) [reply src/sage/combinat/posets/linear_extensions.py]: Also, there is an `x` missing.

**Comment by Sandstorm831** [comment]
@mantepse, done the corrections

**Comment by mantepse** [comment]
Dear @Sandstorm831, please excuse my being pedantic, I noticed another slight simplification of the code.  It is nothing spectacular, but I think it makes the intent clearer.  Warning: we now need to produce `L` as a list, because otherwise we exhaust the iterator.
```py
       if not self:
            return True
        H = self.poset().hasse_diagram()
        L = sources = H.sources()
        linext = []
        for e in self:
            if e not in L:
                return False
            linext.append(e)
            for y in reversed(linext):
                L = [x for x in H.neighbor_out_iterator(y)
                     if x not in linext
                     and all(low in linext for low in H.neighbor_in_iterator(x))]
                if L:
                    break
            else:
                L = sources = [x for x in sources if x not in linext]
        return True
```

**Comment by mantepse** [comment]
In principle we could remove the first two lines, checking whether `self` is empty, too.  However, I guess that this is a valid optimization, we avoid computing the poset, its Hasse diagram and the sources thereof.

**Comment by Sandstorm831** [comment]
@dimpase, sorry if I am missing something, but I want to ask how we avoided computing the poset, hasse diagram and sources.  I can see that we have to calculate all poset, it's hasse diagram and all sources, on the contrary I can't see how this is optimization over the previous code, as I can notice that the check
```
if self[i+1] not in L:
    return False
```
 which was previously present at the bottom, shifted above and L had been initialised earlier.

**Comment by mantepse** [comment]
I wasn't saying that is an optimization, except that the code is now shorter. Instead of checking if 'self[0] not in sources' manually before the loop, it is now done in the loop.

**Comment by Sandstorm831** [comment]
Ohk, sorry @mantepse I misunderstood your point, I will change the code asap.

**Comment by Sandstorm831** [comment]
done! @mantepse 
Also applied as suggested by @dimpase of first two lines, checking whether self is empty.

**Review by mantepse — APPROVED** [review]
LGTM

**Review by dimpase — APPROVED** [review]
lgtm

**Pull request merged**